### ISSUE TITLE
encoding/gocode/gocodec: some functions use deprecated types

### ISSUE THREAD

**Issue by FogDong** [cue-lang/cue#1806]
### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
0.4.3
</pre>

### Does this issue reproduce with the latest release?

Yes.

### What did you do?

Some functions, for example: 

```
// go/encoding/gocode/gocodec/codec.go
// New creates a new Codec for the given instance.
//
// It is safe to use the methods of Codec concurrently as long as the given
// Runtime is not used elsewhere while using Codec. However, only the concurrent
// use of Decode, Validate, and Complete is efficient.
func New(r *cue.Runtime, c *Config) *Codec {
	return &Codec{runtime: value.ConvertToContext(r)}
}
```

`gocodec.New` requires `cue.Runtime` which is deprecated. In the cue code, we can use some internal function to covert runtime like: `value.ConvertToRuntime(cuecontext.New())`, but I can not import the internal package in my go code.

`cue.Instance` is the same, some functions require `cue.Instance`, but I can not use the interval covert `internal.MakeInstance(value).(*cue.Instance)`, is there any workarounds to fix this?

Thanks in advance.

**Comment by myitcv** [comment]
Thanks for raising this and apologies, we are aware of the fact that API is not consistent in this respect (https://github.com/cue-lang/cue/issues/1735, https://github.com/cue-lang/cue/issues/1561, https://github.com/cue-lang/cue/issues/1015). 

For now you can work around this by casting a value of `*cue.Context`:

```
go mod tidy
go run .
cmp stdout stdout.golden

-- go.mod --
module mod.com

go 1.18

require cuelang.org/go v0.4.4-0.20220729051708-0a46a1624353

require (
	github.com/cockroachdb/apd/v2 v2.0.1 // indirect
	github.com/google/uuid v1.2.0 // indirect
	github.com/mpvl/unique v0.0.0-20150818121801-cbe035fff7de // indirect
	github.com/pkg/errors v0.8.1 // indirect
	golang.org/x/net v0.0.0-20220425223048-2871e0cb64e4 // indirect
	golang.org/x/text v0.3.7 // indirect
	gopkg.in/yaml.v3 v3.0.1 // indirect
)
-- main.go --
package main

import (
	"fmt"
	"log"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/encoding/gocode/gocodec"
)

func main() {
	ctx := cuecontext.New()
	g := gocodec.New((*cue.Runtime)(ctx), nil)
	v, err := g.Decode(5)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("v: %v\n", v)
}
-- stdout.golden --
v: 5
```

This works because [`cue.Context`](https://github.com/cue-lang/cue/blob/0a46a1624353da7327d356040a5612251e01659f/cue/context.go#L42) and [`cue.Runtime`](https://github.com/cue-lang/cue/blob/0a46a1624353da7327d356040a5612251e01659f/cue/build.go#L35) have the same underlying type.

**Comment by FogDong** [comment]
Thanks for the reply, is there any other ways to build a cue instance or convert instance from value?
Some functions like `openapi.Gen` still require instance as parameter.

**Comment by myitcv** [comment]
@FogDong - ah that API does need to be updated to use [`cue.InstanceOrValue`](https://pkg.go.dev/cuelang.org/go/cue#InstanceOrValue) (a special transition type). We will submit a fix for that.

**Comment by myitcv** [comment]
FYI @rogpeppe is going to help looking at fixing up these bits of the API.

**Comment by jbcpollak** [comment]
@myitcv is the `cue.InstanceOrValue` change trivial for a newbie to cue to make? I took a look at the code and it didn't seem super obvious what would need to change to make it work, but I'd like to help if possible.

**Comment by myitcv** [comment]
@jbcpollak I'll leave to @rogpeppe to comment. Thanks for your patience!

**Comment by seh** [comment]
This problem came up yesterday [in the "general" channel](https://cuelang.slack.com/archives/CLT3ULF6C/p1681935853668939) of the "CUE" Slack workspace.

**Comment by rogpeppe** [comment]
> @FogDong - ah that API does need to be updated to use [`cue.InstanceOrValue`](https://pkg.go.dev/cuelang.org/go/cue#InstanceOrValue) (a special transition type). We will submit a fix for that.

The API in this case can't be updated to use `InstanceOrValue` AFAICS because the `New` function takes a `*cue.Runtime` not `*cue.Instance`.

I don't see any existing type that's like `InstanceOrValue` but for `RuntimeOrContext` so I'm wondering about adding this to the `cue` package:
```
// ContextOrRuntime is used by some code that needs to accept
// either a *Context or the (deprecated) Runtime type.
type ContextOrRuntime interface {
    runtime() *runtime.Timetime
}

// ToContext returns the context for a value that can be either
// a *Runtime or a *Context.
func ToContext(ctx ContextOrRuntime) *Context {
    return (*Context)(ctx.runtime())
}
```

An alternative might be to do something like:
```
package gocodec
func New[Ctx cue.Runtime|cue.Context](r *Ctx, c *Config) *Codec {
```

**Comment by mvdan** [comment]
@rogpeppe with https://review.gerrithub.io/c/cue-lang/cue/+/556343 merged, are we done for v0.6, or do you plan on doing anything else for this release?

**Comment by rogpeppe** [comment]
@mvdan I went through the API and I think we're probably good. The only possible candidate was `json.NewDecoder` which takes a `*cue.Runtime` argument, but the only reason for passing a non-nil runtime there is to use the `Decode` method which is deprecated, so I think that's OK.

So I'm going to close this as completed.

@FogDong (and any other interested parties) Please comment if you think there's more to be done.
